components/datalogger/datalogger_utils.py

```diff
@@ -230,3 +232,254 @@ def FindAuxControlDataComponentInResourceConfiguration(a_resource_configuration)
+            if val["asset_urn"].startswith("urn:X-topcon:machine"):
+                machine_name = urllib.parse.unquote(val["asset_urn"].split(":")[-1])
+
+            if val["asset_urn"].startswith("urn:X-topcon:device"):
+                device_id = urllib.parse.unquote(val["asset_urn"].split(":")[-1])
+    except KeyError:
+        logging.debug("No machine or device information found.")
+
+    utc_time = datetime.datetime.fromtimestamp(a_replicate["at"]/1000,tz=tz.UTC)
+
+    position_quality = "Unknown"
+
+    try:
+        if aux_control_data_dict["position_quality"] == 0:
+            position_quality = "Unknown"
+        elif aux_control_data_dict["position_quality"] == 1:
+            position_quality = "GPS Float"
+        elif aux_control_data_dict["position_quality"] == 2:
+            position_quality = "RTK Fixed"
+        elif aux_control_data_dict["position_quality"] == 3:
+            position_quality = "mm Enhanced"
+    except:
+        pass
+
+    reverse = "Unknown"
+    position_error_horz = "Unknown"
+    position_error_vert = "Unknown"
+    auto_grade_control = "Unknown"
+
+    try:
+        reverse = aux_control_data_dict["reverse"]
+    except:
+        pass
+    try:
+        position_error_horz = aux_control_data_dict["position_error_horz"]
+    except:
+        pass
+    try:
+        position_error_vert = aux_control_data_dict["position_error_vert"]
+    except:
+        pass
+    try:
+        auto_grade_control = aux_control_data_dict["auto_grade_control"]
+    except:
+        pass
+
+    # Dynamically build the output position_string as a function of the available columns we've been given to write.
+    position_string = ""
+
+    # Output the position columns by iterating over the header list. when we find the headers representing
+    # the point name we're writing, we inject thd data. If we don't find the header names, we add them to
+    # the list and return. They'll then be in the header list the next time we encounter data for that field.
+    for obj in a_object_list:
+        for item in obj["items"]:
+            point_name = item["title"]
+            try:
+                point_name_header_index = a_header_list.index(point_name)
+            except ValueError:
+                # add to list
+                a_header_list.append(point_name)
+                point_name_header_index = a_header_list.index(point_name)
+
+    # Here we populate the value list at the index that matches the POI title for this value in the header list, or None otherwise to correctly space the value list for POIs.
+    # header list, or None otherwise to correctly space the value list for POIs.
+    value_list = [None for _ in range(len(a_header_list))]
+    for obj in a_object_list:
+        for item in obj["items"]:
+            point_name = item["title"]
+            point_name_header_index = a_header_list.index(point_name)
+            value_list[point_name_header_index] = "{}, ".format(item["value"])
```

Comment:
does this result in a trailing , at the end of the value list?

Reply:
Yes. Room for another pass to improve formatting here. Excel hides this thankfully.